doc/source/configuration.rst
============================
Kingbird Configuration Guide
============================

A brief introduction to configure Multisite Kingbird service. Only the
configuration items for Kingbird will be described here. Logging,
messaging, database, keystonemiddleware etc configuration which are
generated from OpenStack OSLO library, will not be described here, for
these configuration items are common to Nova, Cinder, Neutron. So please
refer to corresponding description from Nova or Cinder or Neutron.


Configuration in [DEFAULT]
--------------------------

configuration items for kingbird-api
""""""""""""""""""""""""""""""""""""

bind_host
*********
- default value: *bind_host = 0.0.0.0*
- description: The host IP to bind for kingbird-api service

bind_port
*********
- default value: *bind_port = 8118*
- description: The port to bind for kingbird-api service

api_workers
***********
- default value: *api_workers = 2*
- description: Number of kingbird-api workers

configuration items for kingbird-engine
"""""""""""""""""""""""""""""""""""""""

host
****
- default value: *host = localhost*
- description: The host name kingbird-engine service is running on

workers
*******
- default value: *workers = 1*
- description: Number of kingbird-engine workers

report_interval
***************
- default value: *report_interval = 60*
- description: Seconds between running periodic reporting tasks to
  keep the engine alive in the DB. If the engine doesn't report its
  aliveness to the DB more than two intervals, then the lock accquired
  by the engine will be removed by other engines.

common configuration items for kingbird-api and kingbird-engine
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

use_default_quota_class
***********************
- default value: *use_default_quota_class = true*
- description: Enables or disables use of default quota class with default
  quota, boolean value

Configuration in [kingbird_global_limit]
----------------------------------------

For quota limit, a negative value means unlimited.

configuration items for kingbird-api and kingbird-engine
""""""""""""""""""""""""""""""""""""""""""""""""""""""""

quota_instances
***************
- default value: *quota_instances = 10*
- description: Number of instances allowed per project, integer value.

quota_cores
***********
- default value: *quota_cores = 20*
- description: Number of instance cores allowed per project, integer value.

quota_ram
*********
- default value: *quota_ram = 512*
- description: Megabytes of instance RAM allowed per project, integer value.

quota_metadata_items
********************
- default value: *quota_metadata_items = 128*
- description: Number of metadata items allowed per instance, integer value.

quota_key_pairs
***************
- default value: *quota_key_pairs = 10*
- description: Number of key pairs per user, integer value.

quota_fixed_ips
***************
- default value: *quota_fixed_ips = -1*
- description: Number of fixed IPs allowed per project, this should be at
  least the number of instances allowed, integer value.

quota_security_groups
*********************
- default value: *quota_security_groups = 10*
- description: Number of security groups per project, integer value.

quota_floating_ips
******************
- default value: *quota_floating_ips = 10*
- description: Number of floating IPs allowed per project, integer value.

quota_network
***************
- default value: *quota_network = 10*
- description: Number of networks allowed per project, integer value.

quota_subnet
***************
- default value: *quota_subnet = 10*
- description: Number of subnets allowed per project, integer value.

quota_port
***************
- default value: *quota_port = 50*
- description: Number of ports allowed per project, integer value.

quota_security_group
********************
- default value: *quota_security_group = 10*
- description: Number of security groups allowed per project, integer value.

quota_security_group_rule
*************************
- default value: *quota_security_group_rule = 100*
- description: Number of security group rules allowed per project, integer
  value.

quota_router
************
- default value: *quota_router = 10*
- description: Number of routers allowed per project, integer value.

quota_floatingip
****************
- default value: *quota_floatingip = 50*
- description: Number of floating IPs allowed per project, integer value.

quota_volumes
***************
- default value: *quota_volumes = 10*
- description: Number of volumes allowed per project, integer value.

quota_snapshots
***************
- default value: *quota_snapshots = 10*
- description: Number of snapshots allowed per project, integer value.

quota_gigabytes
***************
- default value: *quota_gigabytes = 1000*
- description: Total amount of storage, in gigabytes, allowed for volumes
  and snapshots per project, integer value.

quota_backups
*************
- default value: *quota_backups = 10*
- description: Number of volume backups allowed per project, integer value.

quota_backup_gigabytes
**********************
- default value: *quota_backup_gigabytes = 1000*
- description: Total amount of storage, in gigabytes, allowed for volume
  backups per project, integer value.

Configuration in [cache]
----------------------------------------

The [cache] section is used by kingbird engine to access the quota
information for Nova, Cinder, Neutron in each region in order to reduce
the KeyStone load while retrieving the endpoint information each time.

configuration items for kingbird-engine
"""""""""""""""""""""""""""""""""""""""

auth_uri
***************
- default value:
- description: Keystone authorization url, for example, http://127.0.0.1:5000/v3.

admin_username
**************
- default value:
- description: Username of admin account, for example, admin.

admin_password
**************
- default value:
- description: Password for admin account, for example, password.

admin_tenant
************
- default value:
- description: Tenant name of admin account, for example, admin.

admin_user_domain_name
**********************
- default value: *admin_user_domain_name = Default*
- description: User domain name of admin account.

admin_project_domain_name
*************************
- default value: *admin_project_domain_name = Default*
- description: Project domain name of admin account.

Configuration in [scheduler]
----------------------------------------

The [scheduler] section is used by kingbird engine to periodically synchronize
and rebalance the quota for each project.

configuration items for kingbird-engine
"""""""""""""""""""""""""""""""""""""""

periodic_enable
***************
- default value: *periodic_enable = True*
- description: Boolean value for enable/disable periodic tasks.

periodic_interval
*****************
- default value: *periodic_interval = 900*
- description: Periodic time interval for automatic quota sync job, unit is
  seconds.

Configuration in [batch]
----------------------------------------

The [batch] section is used by kingbird engine to periodicly synchronize
and rebalance the quota for each project.

batch_size
***************
- default value: *batch_size = 3*
- description: Batch size number of projects will be synced at a time.

Configuration in [locks]
----------------------------------------

The [locks] section is used by kingbird engine to periodically synchronize
and rebalance the quota for each project.

lock_retry_times
****************
- default value: *lock_retry_times = 3*
- description: Number of times trying to grab a lock.

lock_retry_interval
*******************
- default value: *lock_retry_interval =10*
- description: Number of seconds between lock retries.


kingbird/common/config.py
#    Copyright 2016 Ericsson AB
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.

"""
File to store all the configurations
"""
from oslo_config import cfg

global_opts = [
    cfg.BoolOpt('use_default_quota_class',
                default=True,
                help='Enables or disables use of default quota class '
                     'with default quota.'),
    cfg.IntOpt('report_interval',
               default=60,
               help='Seconds between running periodic reporting tasks.'),
]
# Global nova quotas for all projects
nova_quotas = [
    cfg.IntOpt('quota_instances',
               default=10,
               help='Number of instances allowed per project'),
    cfg.IntOpt('quota_cores',
               default=20,
               help='Number of instance cores allowed per project'),
    cfg.IntOpt('quota_ram',
               default=50 * 1024,
               help='Megabytes of instance RAM allowed per project'),
    cfg.IntOpt('quota_floating_ips',
               default=10,
               help='Number of floating IPs allowed per project'),
    cfg.IntOpt('quota_fixed_ips',
               default=-1,
               help='Number of fixed IPs allowed per project (this should be '
                    'at least the number of instances allowed)'),
    cfg.IntOpt('quota_metadata_items',
               default=128,
               help='Number of metadata items allowed per instance'),
    cfg.IntOpt('quota_security_groups',
               default=10,
               help='Number of security groups per project'),
    cfg.IntOpt('quota_key_pairs',
               default=100,
               help='Number of key pairs per user'),
]

# Global neutron quotas for all projects
neutron_quotas = [
    cfg.IntOpt('quota_network',
               default=10,
               help='Number of networks allowed per project. '
                    'A negative value means unlimited.'),
    cfg.IntOpt('quota_subnet',
               default=10,
               help='Number of subnets allowed per project, '
                    'A negative value means unlimited.'),
    cfg.IntOpt('quota_port',
               default=50,
               help='Number of ports allowed per project. '
                    'A negative value means unlimited.'),
    cfg.IntOpt('quota_security_group',
               default=10,
               help='Number of security groups allowed per project. '
                    'A negative value means unlimited.'),
    cfg.IntOpt('quota_security_group_rule',
               default=100,
               help='Number of security rules allowed per project. '
                    'A negative value means unlimited.'),
    cfg.IntOpt('quota_router',
               default=10,
               help='Number of routers allowed per project. '
                    'A negative value means unlimited.'),
    cfg.IntOpt('quota_floatingip',
               default=50,
               help='Number of floating IPs allowed per project. '
                    'A negative value means unlimited.')
]

# Global cinder quotas for all projects
cinder_quotas = [
    cfg.IntOpt('quota_volumes',
               default=10,
               help='Number of volumes allowed per project.'),
    cfg.IntOpt('quota_snapshots',
               default=10,
               help='Number of volume snapshots allowed per project.'),
    cfg.IntOpt('quota_gigabytes',
               default=1000,
               help='Total amount of storage, in gigabytes, allowed '
                    'for volumes and snapshots per project.'),
    cfg.IntOpt('quota_backups', default=10,
               help='Number of volume backups allowed per project.'),
    cfg.IntOpt('quota_backup_gigabytes',
               default=1000,
               help='Total amount of storage, in gigabytes, allowed '
                    'for backups per project.')
]

# OpenStack credentials used for Endpoint Cache
cache_opts = [
    cfg.StrOpt('auth_uri',
               help='Keystone authorization url'),
    cfg.StrOpt('identity_uri',
               help='Keystone service url'),
    cfg.StrOpt('admin_username',
               help='Username of admin account, needed when'
                    ' auto_refresh_endpoint set to True'),
    cfg.StrOpt('admin_password',
               help='Password of admin account, needed when'
                    ' auto_refresh_endpoint set to True'),
    cfg.StrOpt('admin_tenant',
               help='Tenant name of admin account, needed when'
                    ' auto_refresh_endpoint set to True'),
    cfg.StrOpt('admin_user_domain_name',
               default='Default',
               help='User domain name of admin account, needed when'
                    ' auto_refresh_endpoint set to True'),
    cfg.StrOpt('admin_project_domain_name',
               default='Default',
               help='Project domain name of admin account, needed when'
                    ' auto_refresh_endpoint set to True')
]

scheduler_opts = [
    cfg.BoolOpt('periodic_enable',
                default=True,
                help='boolean value for enable/disenable periodic tasks'),
    cfg.IntOpt('periodic_interval',
               default=900,
               help='periodic time interval for automatic quota sync job')
]

common_opts = [
    cfg.IntOpt('workers', default=1,
               help='number of workers'),
    cfg.StrOpt('host',
               default='localhost',
               help='hostname of the machine')
]

scheduler_opt_group = cfg.OptGroup('scheduler',
                                   title='Scheduler options for periodic job')
# The group stores Kingbird global limit for all the projects
default_quota_group = cfg.OptGroup(name='kingbird_global_limit',
                                   title='Global quota limit for all projects')

cache_opt_group = cfg.OptGroup(name='cache',
                               title='OpenStack Credentials')


def list_opts():
    yield default_quota_group.name, nova_quotas
    yield default_quota_group.name, neutron_quotas
    yield default_quota_group.name, cinder_quotas
    yield cache_opt_group.name, cache_opts
    yield scheduler_opt_group.name, scheduler_opts
    yield None, global_opts
    yield None, common_opts


def register_options():
    for group, opts in list_opts():
        cfg.CONF.register_opts(opts, group=group)
